packages/next-swc/crates/core/src/relay.rs

```diff
@@ -0,0 +1,284 @@
+use pathdiff::diff_paths;
+use regex::{Captures, Regex};
+use relay_compiler_common::SourceLocationKey;
+use std::borrow::Cow;
+use std::path::PathBuf;
+
+use relay_compiler::compiler_state::{SourceSet, SourceSetName};
+use relay_compiler::{create_path_for_artifact, FileCategorizer, FileGroup, ProjectConfig};
+use serde::Deserialize;
+use swc_atoms::JsWord;
+use swc_common::errors::HANDLER;
+use swc_common::FileName;
+use swc_ecmascript::ast::*;
+use swc_ecmascript::utils::{quote_ident, ExprFactory};
+use swc_ecmascript::visit::{Fold, FoldWith};
+
+#[derive(Copy, Clone, Debug, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum RelayLanguageConfig {
+    Typescript,
+    Flow,
+}
+
+struct Relay {
+    file_name: FileName,
+    relay_config_for_tests: Option<ProjectConfig>,
+}
+
+fn pull_first_operation_name_from_tpl(tpl: &TaggedTpl) -> Option<String> {
+    tpl.tpl
+        .quasis
+        .iter()
+        .filter_map(|quasis| {
+            let regex = Regex::new(r"(fragment|mutation|query|subscription) (\w+)").unwrap();
+            let capture_group = regex.captures_iter(&quasis.raw.value).next();
+
+            match capture_group {
+                None => None,
+                Some(capture_group) => {
+                    let operation_name: &str = &capture_group[2];
+
+                    Some(operation_name.to_string())
+                }
+            }
+        })
+        .next()
+}
+
+fn build_require_expr_from_path(path: String) -> Expr {
+    Expr::Call(CallExpr {
+        span: Default::default(),
+        callee: quote_ident!("require").as_callee(),
+        args: vec![Lit::Str(Str {
+            span: Default::default(),
+            value: JsWord::from(path),
+            has_escape: false,
+            kind: Default::default(),
+        })
+        .as_arg()],
+        type_args: None,
+    })
+}
+
+impl Fold for Relay {
+    fn fold_expr(&mut self, expr: Expr) -> Expr {
+        let expr = expr.fold_children_with(self);
+
+        match &expr {
+            Expr::TaggedTpl(tpl) => {
+                if let Some(built_expr) = self.build_call_expr_from_tpl(tpl) {
+                    built_expr
+                } else {
+                    expr
+                }
+            }
+            _ => expr,
+        }
+    }
+}
+
+#[derive(Debug)]
+enum BuildRequirePathError<'a> {
+    FileNameNotReal,
+    MultipleSourceSetsFound {
+        source_set_names: Vec<SourceSetName>,
+        path: &'a PathBuf,
+    },
+    ProjectNotFoundForSourceSet {
+        source_set_name: SourceSetName,
+    },
+    FileNotASourceFile,
+    CouldNotCategorize {
+        err: Cow<'static, str>,
+        path: String,
+    },
+}
+
+// This is copied from https://github.com/facebook/relay/blob/main/compiler/crates/relay-compiler/src/build_project/generate_artifacts.rs#L251
+// until the Relay team exposes it for external use.
+fn path_for_artifact(
+    root_dir: &PathBuf,
+    source_path: &PathBuf,
```

Comment:
Same here, `&Path` is better due to deref coercion of rust. You can pass `&PathBuf` as an argument for a function accepts `&Path`, but not vice versa.